src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -33,140 +32,196 @@ internal partial class SyntacticClassificationTaggerProvider
+            // be accessed from both threads, and must be obtained when this lock is held. 
+            //
+            // _lastProcessedRoot is optional and is held onto for languages that support syntax.
+            // it allows computing the root, and changed-spans, in the background so that we can
+            // report a smaller change range, and have the data ready for classifying with GetTags
+            // get called.
+
+            private readonly object _gate = new();
+            private ITextSnapshot? _lastProcessedSnapshot;
+            private Document? _lastProcessedDocument;
+            private SyntaxNode? _lastProcessedRoot;
+
+            // this will cache previous classification information for a span, so that we can avoid
+            // digging into same tree again and again to find exactly same answer
+            private readonly LastLineCache _lastLineCache;
+
             private int _taggerReferenceCount;
 
             public TagComputer(
                 SyntacticClassificationTaggerProvider taggerProvider,
                 ITextBuffer subjectBuffer,
-                IForegroundNotificationService notificationService,
                 IAsynchronousOperationListener asyncListener,
                 ClassificationTypeMap typeMap,
                 TimeSpan diffTimeout)
+                : base(taggerProvider.ThreadingContext, assertIsForeground: false)
             {
+                _taggerProvider = taggerProvider;
                 _subjectBuffer = subjectBuffer;
-                _notificationService = notificationService;
                 _listener = asyncListener;
                 _typeMap = typeMap;
-                _taggerProvider = taggerProvider;
                 _diffTimeout = diffTimeout;
-                _workQueue = new AsynchronousSerialWorkQueue(taggerProvider._threadingContext, asyncListener);
-                _reportChangeCancellationSource = new CancellationTokenSource();
+                _workQueue = new AsyncBatchingWorkQueue<ITextSnapshot>(
+                    TimeSpan.FromMilliseconds(TaggerConstants.NearImmediateDelay),
+                    ProcessChangesAsync,
+                    equalityComparer: null,
+                    asyncListener,
+                    _disposalCancellationSource.Token);
 
-                _lastLineCache = new LastLineCache();
+                _lastLineCache = new LastLineCache(taggerProvider.ThreadingContext);
 
                 _workspaceRegistration = Workspace.GetWorkspaceRegistration(subjectBuffer.AsTextContainer());
                 _workspaceRegistration.WorkspaceChanged += OnWorkspaceRegistrationChanged;
 
                 if (_workspaceRegistration.Workspace != null)
-                {
                     ConnectToWorkspace(_workspaceRegistration.Workspace);
+
+                // Prefer to hear about text changes in the BG so we can use as little 
+                if (_subjectBuffer is ITextBuffer2 textBuffer2)
```

Comment:
@jasonmalinowski @KirillOsenkov are there any cases (liek vs4mac where we might not have an ITextBuffer2)?  If we will always have one, i can simplify thigns by not having the fallback case.

Reply:
VSMac has ITextBuffer2